### ISSUE TITLE
Define types for custom functions

### ISSUE THREAD

**Issue by mainrs** [dhall-lang/dhall-lang#1276]
Not sure if this should be posted here or on https://github.com/dhall-lang/dhall-haskell. 

If I specify my own Dhall-API and provide it through a GitHub repository, and the Haskell runtime has custom functions added to it that Dhall can call. How do I specify those in the Dhall-API? Each time I add them, the LSP throws me errors around my head.

I thought something like this should work, but it's highlighted red:

```dhall
let getAuthor
    : → String
    = API/getAuthor
|```

And `API/getAuthor` would be the function I build into `dhall-haskell`.

**Comment by Gabriel439** [comment]
Yeah, that seems surprising.  If you built a fork of Dhall with a new built-in and built the language server against your fork of Dhall then it should respect the new builtin.  Is it possible that the language server was built against vanilla Dhall and not your fork?

Another thing to check is if the language server auto-completes that same built-in

**Comment by Gabriel439** [comment]
Yeah, this might be possible and it might be possible to upstream into the dhall interpreter, too.  The rough idea I'm thinking of is that the user could provide a record type with expected built-ins, like this:

```dhall
{ API/getAuthor : …, … }
```

… and then pass them in on the command line.  Then the interpreter could add those to the context.

**Comment by mainrs** [comment]
Would this require a change  for the specification? Or is this something implementation-specific? In either case, I can open up appropriate issues if that is OK!

**Comment by Gabriel439** [comment]
This would be specific to the Haskell implementation.  The user experience I had in mind was something like a new `--context ${EXPRESSION}` flag to both the `dhall` and `dhall-lsp-server` executables

**Comment by mainrs** [comment]
I'll close this now, as I've created the followup issue.
